### PR TITLE
Move patch_back_source_tree an action property rather than a rule one

### PR DESCRIPTION
**Pull request by jeremiedimino**
Morally, this is a property of the action rather than the rule.

### CODE REVIEW

File: src/dune_engine/action.ml [hunk 1]
```diff
@@ -292,31 +304,68 @@ module Full = struct
...
+  let adapt_action_for_patch_back_source_tree t =
+    if not t.patch_back_source_tree then
+      t
+    else
+      (* Rules that patch back the source tree cannot go in the shared cache *)
+      let can_go_in_shared_cache = false in
+      (* Rules that patch back the source tree must be sandboxed in copy
+         mode. *)
+      let sandbox = Sandbox_mode.Set.singleton (Some Copy) in
+      { t with can_go_in_shared_cache; sandbox }
```
Comment (snowleopard) [inline src/dune_engine/action.ml:341]: Here we are throwing away the previous `sandbox` setting, and it's not obvious this is the right thing to do (perhaps, we should `inter` them?).

File: src/dune_engine/action.mli [hunk 2]
```diff
@@ -122,19 +122,34 @@ module Full : sig
...
   (** A full action with its environment and list of locks *)
-  type t =
+  type t = private
     { action : action
     ; env : Env.t
     ; locks : Path.t list
     ; can_go_in_shared_cache : bool
     ; sandbox : Sandbox_config.t
+    ; patch_back_source_tree : bool
+          (** Apply all the changes that happend in the sandbox to the source
```
Comment (snowleopard) [inline src/dune_engine/action.mli:132]:
```suggestion
          (** Apply all the changes that happened in the sandbox to the source
```

File: src/dune_engine/action.ml (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_engine/action.ml:343]: intersection seems good. Done

File: src/dune_engine/action.mli (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/action.mli]: Fixed

File: src/dune_rules/dune_file.ml [hunk 3]
```diff
@@ -1686,7 +1706,34 @@ module Rule = struct
...
+           if
+             List.exists (Bindings.to_list deps) ~f:(function
+               | Dep_conf.Sandbox_config _ -> true
+               | _ -> false)
+           then
+             User_error.raise ~loc
+               [ Pp.text
+                   "Rules with (mode patch-back-source-tree) cannot have an \
+                    explicit sandbox configuration has it is implied by (mode \
+                    patch-back-source-tree)."
```
Comment (snowleopard) [inline src/dune_rules/dune_file.ml:1722]: ... implied to be `(sandbox always)` to be more explicit?

**Review by snowleopard — APPROVED** [review]
Looks good!

File: src/dune_rules/dune_file.ml (hunk 3 above)
Comment (jeremiedimino) [reply src/dune_rules/dune_file.ml:1722]: Technically it's `(sandbox copy)`, which is not expressible in a `dune` file. That's why I didn't put it

**Pull request merged**